File: frontend/src/utils/geo.js
export function distanceMetres(lat1, lon1, lat2, lon2) {
  const R = 6371000;
  const toRad = (d) => (d * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  return 2 * R * Math.asin(Math.sqrt(a));
}

export function formatDistance(m) {
  if (m >= 1000) return `${(m / 1000).toFixed(1)} km`;
  return `${Math.round(m)} m`;
}

// Rough ETA assuming a long-distance bus average speed of ~45 km/h.
export function estimateEtaMinutes(m, avgSpeedKmh = 45) {
  const km = m / 1000;
  return Math.round((km / avgSpeedKmh) * 60);
}

File: flutter_app/lib/utils/geo.dart
import 'dart:math';

double distanceMetres(double lat1, double lon1, double lat2, double lon2) {
  const R = 6371000.0;
  final phi1 = lat1 * pi / 180;
  final phi2 = lat2 * pi / 180;
  final deltaPhi = (lat2 - lat1) * pi / 180;
  final deltaLambda = (lon2 - lon1) * pi / 180;

  final a = sin(deltaPhi / 2) * sin(deltaPhi / 2) +
      cos(phi1) * cos(phi2) * sin(deltaLambda / 2) * sin(deltaLambda / 2);

  final c = 2 * atan2(sqrt(a), sqrt(1 - a));

  return R * c;
}

String formatDistance(double metres) {
  if (metres >= 1000) {
    return '${(metres / 1000).toStringAsFixed(1)} km';
  }
  return '${metres.round()} m';
}
